### ISSUE TITLE
Cannot read property 'disallowed_file_patterns' of undefined: Invalid or incomplete plugin settings aren't handled gracefully

### ISSUE THREAD

**Issue by shijistar** [victorporof/Sublime-HTMLPrettify#454]
When I updated to new version, it always crashes when I press ctrl+shift+h. 

The diagnostics is below:
DPI scale: 1.05
startup, version: 3126 linux x64 channel: stable
executable: /opt/sublime_text/sublime_text
working dir: /
packages path: /home/user/.config/sublime-text-3/Packages
state path: /home/user/.config/sublime-text-3/Local
zip path: /opt/sublime_text/Packages
zip path: /home/user/.config/sublime-text-3/Installed Packages
ignored_packages: ["JavaScript Patterns", "Vintage"]
pre session restore time: 0.142178
loading dictionary Packages/Language - English/en_US.dic
startup time: 0.286802
first paint time: 0.367111
first paint time: 0.416362
reloading plugin Default.auto_indent_tag
reloading plugin Default.block
reloading plugin Default.comment
reloading plugin Default.convert_syntax
reloading plugin Default.copy_path
reloading plugin Default.delete_word
reloading plugin Default.detect_indentation
reloading plugin Default.duplicate_line
reloading plugin Default.echo
reloading plugin Default.exec
reloading plugin Default.fold
reloading plugin Default.font
reloading plugin Default.goto_line
reloading plugin Default.history_list
reloading plugin Default.indentation
reloading plugin Default.install_package_control
reloading plugin Default.kill_ring
reloading plugin Default.mark
reloading plugin Default.new_templates
reloading plugin Default.open_context_url
reloading plugin Default.open_in_browser
reloading plugin Default.pane
reloading plugin Default.paragraph
reloading plugin Default.paste_from_history
reloading plugin Default.profile
reloading plugin Default.quick_panel
reloading plugin Default.run_syntax_tests
reloading plugin Default.save_on_focus_lost
reloading plugin Default.scroll
reloading plugin Default.set_unsaved_view_name
reloading plugin Default.settings
reloading plugin Default.show_scope_name
reloading plugin Default.side_bar
reloading plugin Default.sort
reloading plugin Default.swap_line
reloading plugin Default.switch_file
reloading plugin Default.symbol
reloading plugin Default.transform
reloading plugin Default.transpose
reloading plugin Default.trim_trailing_white_space
reloading plugin CSS.css_completions
reloading plugin Diff.diff
reloading plugin HTML.encode_html_entities
reloading plugin HTML.html_completions
reloading plugin 0_package_control_loader.00-package_control
reloading plugin Better Completion.main
reloading plugin CSS Extended Completions.cache
reloading plugin CSS Extended Completions.commands
reloading plugin CSS Extended Completions.completions
reloading plugin CSS Extended Completions.css_style_completions
reloading plugin CSS Extended Completions.extended_css_completions
reloading plugin CSS Extended Completions.location
reloading plugin CSS Extended Completions.project
reloading plugin CSS Extended Completions.settings
reloading plugin CSS Extended Completions.style_parser
reloading plugin DocBlockr.jsdocs
reloading plugin Git.git_commands
reloading plugin Package Control.1_reloader
reloading plugin Package Control.2_bootstrap
reloading plugin Package Control.Package Control
reloading plugin HTML-CSS-JS Prettify.HTMLPrettify
plugins loaded
Package Control: Skipping automatic upgrade, last run at 2017-07-19 12:25:13, next run at 2017-07-19 13:25:13 or after
Node.js encountered a runtime error: Unhandled promise rejection TypeError: Cannot read property 'disallowed_file_patterns' of undefined
    at isDisallowedFilePattern (/home/user/.config/sublime-text-3/Packages/HTML-CSS-JS Prettify/build/js-transpiled/utils/fileUtils.js:38:80)
    at isCSS (/home/user/.config/sublime-text-3/Packages/HTML-CSS-JS Prettify/build/js-transpiled/utils/fileUtils.js:135:7)
    at _callee$ (/home/user/.config/sublime-text-3/Packages/HTML-CSS-JS Prettify/build/js-transpiled/main.js:85:38)
    at tryCatch (/home/user/.config/sublime-text-3/Packages/HTML-CSS-JS Prettify/build/js-transpiled/node_modules/regenerator-runtime/runtime.js:65:40)
    at GeneratorFunctionPrototype.invoke [as _invoke] (/home/user/.config/sublime-text-3/Packages/HTML-CSS-JS Prettify/build/js-transpiled/node_modules/regenerator-runtime/runtime.js:303:22)
    at GeneratorFunctionPrototype.prototype.(anonymous function) [as next] (/home/user/.config/sublime-text-3/Packages/HTML-CSS-JS Prettify/build/js-transpiled/node_modules/regenerator-runtime/runtime.js:117:21)
    at step (/home/user/.config/sublime-text-3/Packages/HTML-CSS-JS Prettify/build/js-transpiled/node_modules/babel-runtime/helpers/asyncToGenerator.js:17:30)
    at /home/user/.config/sublime-text-3/Packages/HTML-CSS-JS Prettify/build/js-transpiled/node_modules/babel-runtime/helpers/asyncToGenerator.js:28:13

### HTMLPrettify diagnostics begin ###
[HTMLPrettify] Using editor text temp file: True
[HTMLPrettify] Global file rules: {"html": {"allowed_file_extensions": ["htm", "html", "xhtml", "shtml", "xml", "svg", "hbs"], "allowed_file_syntaxes": ["html", "xml"], "disallowed_file_patterns": []}}
[HTMLPrettify] Respecting .editorconfig files: True
[HTMLPrettify] Editor file syntax: Packages/JavaScriptNext - ES6 Syntax/JavaScriptNext.tmLanguage
[HTMLPrettify] Editor indent size: ?
[HTMLPrettify] Editor indent with tabs: ?
[HTMLPrettify] Editor text file path: /home/user/.config/sublime-text-3/Packages/HTML-CSS-JS Prettify/4ac1fbcf-18da-47e5-b6c5-0f9bb6bce136
[HTMLPrettify] Editor text file contents: null
[HTMLPrettify] Original file path: /home/user/Documents/udesk_case_web/app/initializers/init-component-summer-note.js
[HTMLPrettify] Config extra lookup paths: /home/user/.config/sublime-text-3/Packages/User,/home/user/.config/sublime-text-3/Packages/HTML-CSS-JS Prettify
[HTMLPrettify] Computed extra lookup paths for .jsbeautifyrc: ["/home/user/Documents/udesk_case_web/app/initializers","/home/user/Documents/udesk_case_web/app","/home/user/Documents/udesk_case_web","/home/user/Documents","/home/user","/home","/home/user","/home/user/.config/sublime-text-3/Packages/User","/home/user/.config/sublime-text-3/Packages/HTML-CSS-JS Prettify"]
[HTMLPrettify] Attempting to parse file: /home/user/.config/sublime-text-3/Packages/HTML-CSS-JS Prettify/.jsbeautifyrc.defaults.json
[HTMLPrettify] Attempting to parse file: /home/user/.config/sublime-text-3/Packages/User/.jsbeautifyrc
[HTMLPrettify] Attempting to parse file: /home/user/Documents/udesk_case_web/.editorconfig
[HTMLPrettify] Computed prettify options: {"html":{"end_with_newline":true,"eol":"\n","indent_level":0,"indent_char":" ","indent_size":"4","indent_with_tabs":false,"max_preserve_newlines":0,"preserve_newlines":true,"wrap_line_length":0,"indent_handlebars":true,"indent_inner_html":false,"indent_scripts":"keep","unformatted":["img","code","pre","sub","sup","em","b","i","u","strike","big","pre","h1"],"css":{"newline_between_rules":true,"selector_separator":" ","selector_separator_newline":true},"js":{"brace_style":"collapse-preserve-inline","break_chained_methods":false,"comma_first":false,"e4x":false,"jslint_happy":false,"keep_array_indentation":false,"keep_function_indentation":false,"operator_position":"before-newline","space_after_anon_function":true,"space_before_conditional":true,"space_in_empty_paren":false,"space_in_paren":false,"unescape_strings":false},"end_of_line":"lf","charset":"utf-8","trim_trailing_whitespace":true},"css":{"end_with_newline":true,"eol":"\n","indent_level":0,"indent_char":" ","indent_size":"4","indent_with_tabs":false,"max_preserve_newlines":0,"preserve_newlines":true,"wrap_line_length":0,"newline_between_rules":true,"selector_separator":" ","selector_separator_newline":true,"end_of_line":"lf","charset":"utf-8","trim_trailing_whitespace":true},"js":{"end_with_newline":true,"eol":"\n","indent_level":0,"indent_char":" ","indent_size":"4","indent_with_tabs":false,"max_preserve_newlines":0,"preserve_newlines":true,"wrap_line_length":0,"brace_style":"collapse-preserve-inline","break_chained_methods":false,"comma_first":false,"e4x":false,"jslint_happy":false,"keep_array_indentation":false,"keep_function_indentation":false,"operator_position":"before-newline","space_after_anon_function":true,"space_before_conditional":true,"space_in_empty_paren":false,"space_in_paren":false,"unescape_strings":false,"end_of_line":"lf","charset":"utf-8","trim_trailing_whitespace":true},"json":{"end_with_newline":true,"eol":"\n","indent_level":0,"indent_char":" ","indent_size":"4","indent_with_tabs":false,"max_preserve_newlines":0,"preserve_newlines":true,"wrap_line_length":0,"brace_style":"expand","keep_array_indentation":false,"unescape_strings":false,"end_of_line":"lf","charset":"utf-8","trim_trailing_whitespace":true}}

**Comment by victorporof** [comment]
Thank you, this looks like a plugin bug. Does it still haopen in the latest version? 

I'll take a look today.

**Comment by shijistar** [comment]
Yes, it's the latest version. I think I find the problem and have fixed it. I thought the `sublime-text-3/Packages/User/HTMLPrettify.sublime-settings` would deeply merged into `sublime-text-3/Packages/HTML-CSS-JS Prettify/HTMLPrettify.sublime-settings`, so I only write the as less as possible configuration which is different as default values. Below is the file content:
```javascript
{
    // Simply using `node` without specifying a path sometimes doesn't work :(
    // https://github.com/victorporof/Sublime-HTMLPrettify#oh-noez-command-not-found
    // http://nodejs.org/#download
    "node_path":
    {
        "windows": "C:/Program Files/nodejs/node.exe",
        "linux": "/home/user/.nvm/versions/node/v6.9.5/bin/node",
        "osx": "/usr/local/bin/node"
    },

    // Automatically format when a file is saved.
    "format_on_save": false,

    // Settings determining which files are allowed to be prettified.
    "global_file_rules":
    {
        "html":
        {
            "allowed_file_extensions": ["htm", "html", "xhtml", "shtml", "xml", "svg", "hbs"],
            "allowed_file_syntaxes": ["html", "xml"],
            "disallowed_file_patterns": []
        }
    }
}
```
But actually it's not. I guess it just does shadow copy, so `global_file_rules` only has `html` field but other fields such as `css` are all disappeared. So that's why the error message is: Cannot read property 'disallowed_file_patterns' of undefined.
When I copy the whole default configuration file and make the changes I need, then the error disappears.

**Comment by victorporof** [comment]
I see. Well, this is sublime's default behavior and doesn't really have anything to do with the plugin, but we should either fail more gracefully or deal with this better.

**Comment by victorporof** [comment]
For anyone stumbling upon this issue, `global_file_rules` needs to contain all necessary keys, so `html`, `css`, `js`, `json`.

**Comment by victorporof** [comment]
Fixed in 2.1.1
